### ISSUE TITLE
Agent choking our applications - Reentrant locks perhaps the culprit

### ISSUE THREAD

**Issue by robanoz** [microsoft/ApplicationInsights-Java#2056]
We are facing intermittent issues where our applications randomly stop responding after a while. Based on thread dumps, they all report the same, having a large number of threads waiting for the same reentrant lock.

A bit of background: We now have more than 200+ containerized application in AKS, reporting to the same log analytics workspace using agent. While the issue was first reported back in November, we initially believed it could be related to us upgrading to 3.2.3. But downgrading to 3.2.2  did not solve it, it perhaps made it somewhat worse. Then we assumed it may be related to the preview sampling feature introduced when we first upgrade, where we filtered out health probe URLs spamming our logs and later some noisy SQLs. Today, we confirmed the same was happening again after an upgrade last evening, on 3.2.3 and without anything set for sampling in the applicationInsights.json.

Observations:
- When the application stops to respond, we see an unusual amount of threads in waiting state (e.g. 150 vs. 20 in normal)
- The thread dump (attached) points to a ReentrantLock where these threads are waiting for the same lock
- The lock is within the open telemetry API (it seems), and "blocks" anything related to agent instrumentation (logging, SQL etc.)
- A log statement we typically see in all these cases, as last before we have to restart is this:
`OpenJDK 64-Bit Server VM warning: Potentially dangerous stack overflow in ReservedStackAccess annotated method java.util.concurrent.locks.ReentrantLock$Sync.nonfairTryAcquire(I)Z [1]`
- Could this be related to the [bugfix in 1.7.1](https://github.com/open-telemetry/opentelemetry-java/issues/3813)? I see that there is a leap to 1.10 in the latest beta, but we would rather not experiment with beta in production.
- When this happen, the only remedy is to restart the application

This is a snippet from thread dump, that shows one Http thread waiting on the lock:
`at jdk.internal.misc.Unsafe.park(java.base@11.0.6/Native Method)
at java.util.concurrent.locks.LockSupport.park(java.base@11.0.6/LockSupport.java:194)
at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(java.base@11.0.6/AbstractQueuedSynchronizer.java:885)
at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(java.base@11.0.6/AbstractQueuedSynchronizer.java:917)
at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@11.0.6/AbstractQueuedSynchronizer.java:1240)
at java.util.concurrent.locks.ReentrantLock.lock(java.base@11.0.6/ReentrantLock.java:267)
at java.util.concurrent.ArrayBlockingQueue.offer(java.base@11.0.6/ArrayBlockingQueue.java:341)
at io.opentelemetry.sdk.trace.export.BatchSpanProcessor$Worker.addSpan(BatchSpanProcessor.java:205)
at io.opentelemetry.sdk.trace.export.BatchSpanProcessor$Worker.access$100(BatchSpanProcessor.java:125)
at io.opentelemetry.sdk.trace.export.BatchSpanProcessor.onEnd(BatchSpanProcessor.java:97)
at io.opentelemetry.sdk.trace.MultiSpanProcessor.onEnd(MultiSpanProcessor.java:52)
at io.opentelemetry.sdk.trace.RecordEventsReadableSpan.endInternal(RecordEventsReadableSpan.java:460)
at io.opentelemetry.sdk.trace.RecordEventsReadableSpan.end(RecordEventsReadableSpan.java:440)
at io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.Instrumenter.end(Instrumenter.java:213)
at com.microsoft.sqlserver.jdbc.SQLServerStatement.execute(SQLServerStatement.java:746)`

We have a practice now of pulling out thread dumps, and they all look the same - vs. a normal dump with no signs of this. See attached full thread dump.

### System information
Please provide the following information:
 - OpenJDK 64-Bit Server VM (11.0)
 - Agent version: 3.2.2-3.2.3, currently testing 3.2.4
 - OS type and version: Debian GNU/Linux 10 (buster)
 - Application Server type and version (if applicable): Tomcat  9.0
 - Using spring-boot? No
 - Additional relevant libraries (with version, if applicable): N/A

applicationinsights.json:
`{
  "customDimensions": {
  },
  "instrumentation": {
    "logging": {
      "level": "ALL"
    },
    "micrometer": {
      "enabled": false
    }
  },
  "selfDiagnostics": {
    "destination": "file+console",
    "level": "WARN",
    "file": {
      "path": "applicationinsights.log",
      "maxSizeMb": 5,
      "maxHistory": 1
    }
  }
}`

### Logs
The reports 1-2 cases of truncated SQL statements, but other than that, is often blank. 
[threadDump.txt](https://github.com/microsoft/ApplicationInsights-Java/files/7888914/threadDump.txt)

**Comment by trask** [comment]
hi @robanoz,

which minor version of OpenJDK 11 are you using?

and do you see this message before or after you start seeing the bottlenecking on ReentrantLock?
   `OpenJDK 64-Bit Server VM warning: Potentially dangerous stack overflow in ReservedStackAccess annotated method java.util.concurrent.locks.ReentrantLock$Sync.nonfairTryAcquire(I)Z`

**Comment by robanoz** [comment]
> hi @robanoz,
> 
> which minor version of OpenJDK 11 are you using?
> 
> and do you see this message before or after you start seeing the bottlenecking on ReentrantLock? `OpenJDK 64-Bit Server VM warning: Potentially dangerous stack overflow in ReservedStackAccess annotated method java.util.concurrent.locks.ReentrantLock$Sync.nonfairTryAcquire(I)Z`

Hi @trask , we are using 11.0.6+10. It is a bit hard to tell exactly when the message appears. But what we know, is that it is more or less the last thing the application logs. I would suspect that is since even the logging is being blocked, as we have seen from the thread dump.

**Comment by trask** [comment]
this looks like a JVM bug:

```
OpenJDK 64-Bit Server VM warning: Potentially dangerous stack overflow in ReservedStackAccess annotated method java.util.concurrent.locks.ReentrantLock$Sync.nonfairTryAcquire(I)Z
```

and my guess is that is leading to the lock not being released, which is then why all the threads block on getting that lock.

can you try updating to the latest Java 11 patch release to see if we get lucky and it's fixed there?

the warnings above (after updating to 3.2.4) don't particularly look related to the thread blocking that you're seeing

**Comment by trask** [comment]
btw, related to the `Queue is full` warning you are seeing above:

```
WARN c.m.a.a.i.t.BatchSpanProcessor - In the last 5 minutes, the following operation has failed 1325 times (out of 9916): Queuing span: ** Queue is full (1325 times)
```

we found that a common culprit of this is due to bursty metric recording which occurs once a minute, so we have split out metrics into separate queue and increased the size of that queue. this will be available in next release, or now in latest SNAPSHOT: https://github.com/microsoft/ApplicationInsights-Java/suites/4997129948/artifacts/146434756

**Comment by robanoz** [comment]
We did the following which seems to do the trick:

- Updated to latest Java 11 patch release as suggested
- Changed sampling to 50%

We have now run almost 3 days without a single lock. What we would like to do now, is remove the overall sampling (since it is blinding us for especially logs). This will also confirm whether or not is was a Java 11 issue from an earlier patch release.

I suggest we keep the issue open until we have confirmed some time the coming week 😉
